src/metricscollector/collector/collector_test.go

```diff
@@ -18,67 +19,104 @@ import (
 var _ = Describe("Collector", func() {
 
 	var (
-		cfc       *fakes.FakeCfClient
-		noaa      *fakes.FakeNoaaConsumer
-		database  *fakes.FakeDB
-		coll      *Collector
-		duration  time.Duration
-		startTime time.Time
-		pollings  *[3][3]bool
+		cfc      *fakes.FakeCfClient
+		noaa     *fakes.FakeNoaaConsumer
+		database *fakes.FakeDB
+		coll     *Collector
+		fclock   *fakeclock.FakeClock
+		buffer   *gbytes.Buffer
 	)
 
 	BeforeEach(func() {
+
 		cfc = &fakes.FakeCfClient{}
 		noaa = &fakes.FakeNoaaConsumer{}
 		database = &fakes.FakeDB{}
+
+		conf := &config.CollectorConfig{
+			PollInterval:    TestPollInterval,
+			RefreshInterval: TestRefreshInterval,
+		}
+
 		logger := lager.NewLogger("collector-test")
-		conf := &config.CollectorConfig{RefreshInterval: TestRefreshInterval, PollInterval: TestPollInterval}
-		coll = NewCollector(logger, conf, cfc, noaa, database)
+		buffer = gbytes.NewBuffer()
+		logger.RegisterSink(lager.NewWriterSink(buffer, lager.ERROR))
+
+		fclock = fakeclock.NewFakeClock(time.Now())
+		coll = NewCollector(conf, logger, cfc, noaa, database, fclock)
 	})
 
 	Describe("Start", func() {
 		JustBeforeEach(func() {
 			coll.Start()
-			time.Sleep(duration)
+		})
+
+		AfterEach(func() {
 			coll.Stop()
 		})
 
 		Context("when getting apps from policy database succeeds", func() {
 			BeforeEach(func() {
-				pollings = &[3][3]bool{}
-				startTime = time.Now()
-
-				noaa.ContainerMetricsStub = func(appid string, token string) ([]*events.ContainerMetric, error) {
-					computePoolings(appid, startTime, pollings)
+				noaa.ContainerMetricsStub = func(appId string, token string) ([]*events.ContainerMetric, error) {
+					switch appId {
+					case "app-id-1":
+						pollings[0] = true
+					case "app-id-2":
+						pollings[1] = true
+					case "app-id-3":
+						pollings[2] = true
+					}
 					return []*events.ContainerMetric{}, nil
 				}
 			})
 
 			Context("when no apps in policy database", func() {
 				BeforeEach(func() {
-					duration = 2500 * time.Millisecond
 					database.GetAppIdsReturns(make(map[string]bool), nil)
 				})
 
 				It("does not poll anything", func() {
-					Expect(noaa.ContainerMetricsCallCount()).To(BeZero())
+					fclock.Increment(TestRefreshInterval * time.Second)
+					Eventually(database.GetAppIdsCallCount).Should(Equal(1))
```

Comment:
We are confusing 2 things here, polling and refreshing.
The collector only knows about one thing, refreshing the set of pollers.
If you want to test whether pollers are created, started and stopped, which would be a good thing, then I would allow a function to be passed in that did the creation of a AppPoller interface.